Plutarch/Internal/Term.hs

```diff
@@ -102,9 +95,34 @@ import UntypedPlutusCore qualified as UPLC
  though the name is relative to the current level.
 -}
 
-data HoistedTerm = HoistedTerm (Digest Blake2b_160) RawTerm
+data HoistedTerm = HoistedTerm {htHash :: Int, htRawTerm :: RawTerm}
   deriving stock (Show)
 
+{- | Hoisted term carry their own non-cryptographic hash, making it incredibly
+cheap to hold these in hashmaps.
+-}
+instance Hashable HoistedTerm where
+  hashWithSalt = defaultHashWithSalt
+  {-# INLINE hashWithSalt #-}
+
+  -- The instance uses that this is basically 'Hashed a'.
+  hash = htHash
+  {-# INLINE hash #-}
+
+{- | Equality of hoisted terms is first checked via their non-cryptographic
+hash, then via term equality. Inequality, which should happen much more
+often is thus cheap.
+
+Note that we could just as well derive this instance...
+-}
+instance Eq HoistedTerm where
+  l == r =
+    htHash l == htHash r
+      && htRawTerm l == htRawTerm r
+  {-# INLINE (==) #-}
+
+type UTerm = UPLC.Term UPLC.DeBruijn UPLC.DefaultUni UPLC.DefaultFun ()
```

Comment:
I'd rather keep this expanded in places we need it. Yes, it's longer, but I'd rather _know_ that we're referring to some horrid upstream pile and not potentially a type wrapper which I have to retain additional context about.